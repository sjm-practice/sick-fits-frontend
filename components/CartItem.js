import React from "react";
import styled from "styled-components";
import formatMoney from "../lib/formatMoney";
import RemoveFromCart from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => {
  const { item, quantity, id } = cartItem;

  if (!item) {
    return (
      <CartItemStyles>
        <p>This item has been removed.</p>
        <RemoveFromCart id={cartItem.id} />
      </CartItemStyles>
    );
  }

  const { image, title, price } = item;

  return (
    <CartItemStyles>
      <img width="100" src={image} alt={title} />
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(price * quantity)}
          {` - `}
          <em>
            {quantity} &times; {formatMoney(price)} each
          </em>
        </p>
      </div>
      <RemoveFromCart id={id} />
    </CartItemStyles>
  );
};

export default CartItem;
