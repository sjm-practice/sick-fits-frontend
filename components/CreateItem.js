import React, { Component } from "react";
import { Mutation } from "react-apollo";
import Router from "next/router";
import gql from "graphql-tag";
import Form from "./styles/Form";
import ErrorMessage from "./ErrorMessage";

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: "Sample",
    description: "Sample Description",
    image: "",
    largeImage: "large-sample.jpg", // eslint-disable-line react/no-unused-state
    price: 1000,
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    const val = type === "number" ? parseFloat(value) : value;

    this.setState({ [name]: val });
  };

  handleSubmit = async (e, createItem) => {
    e.preventDefault();
    const res = await createItem();
    console.log(res);
    Router.push({
      pathname: "/item",
      query: { id: res.data.createItem.id },
    });
  };

  uploadFile = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "sickfits");

    const res = await fetch("https://api.cloudinary.com/v1_1/smarsh/image/upload", {
      method: "POST",
      body: data,
    });
    const file = await res.json();
    console.log(file);
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url, // eslint-disable-line react/no-unused-state
    });
  };

  render() {
    const { title, price, description, image } = this.state;

    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form onSubmit={e => this.handleSubmit(e, createItem)}>
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="image">
                Image
                <input
                  type="file"
                  id="image"
                  name="image"
                  required
                  placeholder="Upload An Image"
                  onChange={this.uploadFile}
                />
                {image && <img width="200" src={image} alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  required
                  value={price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter A Description"
                  required
                  value={description}
                  onChange={this.handleChange}
                />
              </label>
            </fieldset>
            <button type="submit">Submit</button>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
