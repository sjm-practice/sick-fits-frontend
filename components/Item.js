import React, { Component } from "react";
import PropTypes from "prop-types";
import Link from "next/link";
import Title from "./styles/Title";
import ItemStyles from "./styles/ItemStyles";
import PriceTag from "./styles/PriceTag";

class Item extends Component {
  render() {
    const { item } = this.props;

    return (
      <ItemStyles>
        <Title>
          <Link
            href={{
              pathname: "/item",
              query: { id: item.id },
            }}
          >
            <a>{item.title}</a>
          </Link>
        </Title>
        <div>
          <PriceTag>{item.price}</PriceTag>
        </div>
      </ItemStyles>
    );
  }
}

Item.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string,
    image: PropTypes.string,
    largeImage: PropTypes.string,
    price: PropTypes.number,
  }).isRequired,
};

export default Item;
