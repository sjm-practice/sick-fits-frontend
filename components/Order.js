import React, { Component } from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import { format } from "date-fns";
import Head from "next/head";
import gql from "graphql-tag";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";
import OrderStyles from "./styles/OrderStyles";

export const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`;

class Order extends Component {
  render() {
    const { id: orderId } = this.props;
    return (
      <Query query={SINGLE_ORDER_QUERY} variables={{ id: orderId }}>
        {({ data, error, loading }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;

          const {
            order: { id, charge, createdAt, total, items },
          } = data;

          return (
            <OrderStyles data-test="order">
              <Head>
                <title>Sick Fits! - Order {id}</title>
              </Head>
              <p>
                <span>Order ID:</span>
                <span>{id}</span>
              </p>
              <p>
                <span>Charge:</span>
                <span>{charge}</span>
              </p>
              <p>
                <span>Date:</span>
                <span>{format(createdAt, "MMMM d, YYYY h:mm a")}</span>
              </p>
              <p>
                <span>Total:</span>
                <span>{formatMoney(total)}</span>
              </p>
              <p>
                <span>Item Count:</span>
                <span>{items.length}</span>
              </p>
              <div className="items">
                {items.map(item => (
                  <div className="order-item" key={item.id}>
                    <img src={item.image} alt={item.title} />
                    <div className="item-details">
                      <h2>{item.title}</h2>
                      <p>Qty: {item.quantity}</p>
                      <p>Each: {formatMoney(item.price)}</p>
                      <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
                      <p>Description: {item.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </OrderStyles>
          );
        }}
      </Query>
    );
  }
}

Order.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Order;
