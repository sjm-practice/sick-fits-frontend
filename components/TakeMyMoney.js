import React, { Component } from "react";
import PropTypes from "prop-types";
import Router from "next/router";
import NProgress from "nprogress";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import StripeCheckout from "react-stripe-checkout";
import User, { CURRENT_USER_QUERY } from "./User";
import calcTotalPrice from "../lib/calcTotalPrice";

export const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

function totalItems(cart) {
  return cart.reduce((tally, { quantity }) => tally + quantity, 0);
}

class TakeMyMoney extends Component {
  onToken = async (res, createOrder) => {
    NProgress.start();

    // manually call the mutation once we have the stripe token
    const order = await createOrder({
      variables: {
        token: res.id,
      },
    }).catch(err => alert(err.message));

    Router.push({
      pathname: "/order",
      query: { id: order.data.createOrder.id },
    });
  };

  render() {
    const { children } = this.props;

    return (
      <User>
        {({ data: { me }, loading }) => {
          if (loading) return null;
          const { cart } = me;
          const image =
            cart.length && cart[0].item && cart[0].item.image ? cart[0].item.image : "";

          return (
            <Mutation
              mutation={CREATE_ORDER_MUTATION}
              refetchQueries={[{ query: CURRENT_USER_QUERY }]}
            >
              {createOrder => (
                <StripeCheckout
                  amount={calcTotalPrice(cart)}
                  name="Sick Fits"
                  description={`Order of ${totalItems(cart)} items!`}
                  image={image}
                  stripeKey="pk_test_XmebB427QkON6HMeQq3lAn49"
                  currency="USD"
                  email={me.email}
                  token={res => this.onToken(res, createOrder)}
                >
                  {children}
                </StripeCheckout>
              )}
            </Mutation>
          );
        }}
      </User>
    );
  }
}

TakeMyMoney.propTypes = {
  children: PropTypes.node.isRequired,
};

export default TakeMyMoney;
