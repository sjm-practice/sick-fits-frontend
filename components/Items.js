import React, { Component } from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import Item from "./Item";
import Pagination from "./Pagination";

export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY {
    items {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const Center = styled.div`
  text-align: center;
`;

const ItemsList = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${props => props.theme.maxWidth};
  margin: 0 auto;
`;

class Items extends Component {
  static propTypes = {
    page: PropTypes.number.isRequired,
  };

  render() {
    const { page } = this.props;
    return (
      <Center>
        <Pagination page={page} />
        <Query query={ALL_ITEMS_QUERY}>
          {({ data, error, loading }) => {
            if (loading) return <p>Loading...</p>;
            if (error) return <p>Error: {error.message}</p>;
            return (
              <ItemsList>
                {data.items.map(item => (
                  <Item key={item.id} item={item} />
                ))}
              </ItemsList>
            );
          }}
        </Query>
        <Pagination />
      </Center>
    );
  }
}

export default Items;
