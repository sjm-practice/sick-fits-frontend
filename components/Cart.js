import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import User from "./User";
import CartItem from "./CartItem";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";
import TakeMyMoney from "./TakeMyMoney";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";

export const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const Cart = () => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const {
        data: { me },
      } = user;
      if (!me) return null;

      const { name, cart } = me;
      const {
        data: { cartOpen },
      } = localState;

      return (
        <CartStyles open={cartOpen}>
          <header>
            <CloseButton onClick={toggleCart} title="close">
              &times;
            </CloseButton>
            <Supreme>{name}&#39;s Cart</Supreme>
            <p>
              You have {cart.length} product{cart.length === 1 ? "" : "s"} in your cart.
            </p>
          </header>

          <ul>
            {cart.map(cartItem => (
              <CartItem key={cartItem.id} cartItem={cartItem} />
            ))}
          </ul>
          <footer>
            <p>{formatMoney(calcTotalPrice(cart))}</p>
            {cart.length > 0 && (
              <TakeMyMoney>
                <SickButton>Checkout</SickButton>
              </TakeMyMoney>
            )}
          </footer>
        </CartStyles>
      );
    }}
  </Composed>
);

export default Cart;
