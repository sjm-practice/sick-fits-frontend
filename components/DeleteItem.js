import React, { Component } from "react";
import PropTypes from "prop-types";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  handleDelete = deleteItem => {
    // eslint-disable-next-line no-alert
    if (confirm("Are you sure you want to delete this item?")) {
      deleteItem().catch(err => alert(err.message));
    }
  };

  update = (cache, payload) => {
    // manually update client cache, to match server
    // 1. read the cache for the items we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // 2. filter deleted item out of cache
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    // 3. put the items back in cache
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  };

  render() {
    const { children, id } = this.props;

    return (
      <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id }} update={this.update}>
        {(deleteItem, { error }) => (
          <button type="button" onClick={() => this.handleDelete(deleteItem)}>
            {children}
          </button>
        )}
      </Mutation>
    );
  }
}

DeleteItem.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
};

export default DeleteItem;
