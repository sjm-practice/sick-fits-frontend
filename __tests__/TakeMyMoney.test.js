import React from "react";
import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import Router from "next/router";
import NProgress from "nprogress";
import { MockedProvider } from "react-apollo/test-utils";
import TakeMyMoney from "../components/TakeMyMoney";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import { CURRENT_USER_QUERY } from "../components/User";

Router.router = { push() {} };

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem()],
        },
      },
    },
  },
];

describe("<TakeMyMoney />", () => {
  it("should render and match snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney>
          <div />
        </TakeMyMoney>
      </MockedProvider>,
    );
    await wait();
    wrapper.update();

    expect(toJSON(wrapper.find("ReactStripeCheckout"))).toMatchSnapshot();
  });

  it("should create an order onToken", async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: "xyz789",
        },
      },
    });

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney>
          <div />
        </TakeMyMoney>
      </MockedProvider>,
    );
    const component = wrapper.find("TakeMyMoney").instance();
    // manually call that onToken method
    component.onToken({ id: "abc123" }, createOrderMock);
    expect(createOrderMock).toHaveBeenCalledWith({ variables: { token: "abc123" } });
  });

  it("should start NProgress progress bar", async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: "xyz789",
        },
      },
    });

    NProgress.start = jest.fn();

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney>
          <div />
        </TakeMyMoney>
      </MockedProvider>,
    );
    await wait();
    wrapper.update();

    const component = wrapper.find("TakeMyMoney").instance();
    // manually call that onToken method
    component.onToken({ id: "abc123" }, createOrderMock);
    expect(NProgress.start).toHaveBeenCalledWith();
  });

  it("should route to the order page when completed", async () => {
    const createOrderMock = jest.fn().mockResolvedValue({
      data: {
        createOrder: {
          id: "xyz789",
        },
      },
    });

    Router.router.push = jest.fn();

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <TakeMyMoney>
          <div />
        </TakeMyMoney>
      </MockedProvider>,
    );
    await wait();
    wrapper.update();

    const component = wrapper.find("TakeMyMoney").instance();
    // manually call that onToken method
    component.onToken({ id: "abc123" }, createOrderMock);
    await wait();
    wrapper.update();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: "/order",
      query: { id: "xyz789" },
    });
  });
});
