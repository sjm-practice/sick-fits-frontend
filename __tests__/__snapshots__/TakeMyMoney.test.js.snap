// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TakeMyMoney /> should render and match snapshot 1`] = `
<ReactStripeCheckout
  ComponentClass="span"
  amount={15000}
  className="StripeCheckout"
  currency="USD"
  description="Order of 3 items!"
  email="Delmer.Smith@yahoo.com"
  image="dog-small.jpg"
  label="Pay With Card"
  locale="auto"
  name="Sick Fits"
  reconfigureOnUpdate={false}
  stripeKey="pk_test_XmebB427QkON6HMeQq3lAn49"
  token={[Function]}
  triggerEvent="onClick"
>
  <span
    onClick={[Function]}
  >
    <div />
  </span>
</ReactStripeCheckout>
`;
